src/mscorlib/corefx/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs

```
@@ -0,0 +1,326 @@
+                {
+                    if ((arr = stacks[index].TryPop()) != null) return arr;
+                    if (++index == stacks.Length) index = 0;
+                }
+                return null;
+            }
+        }
+
+        /// <summary>Provides a simple stack of arrays, protected by a lock.</summary>
+        private sealed class LockedStack
+        {
+            private readonly T[][] _arrays = new T[MaxBuffersPerArraySizePerCore][];
+            private int _count;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public bool TryPush(T[] array)
+            {
+                bool enqueued = false;
+                MonitorEnterWithProcNumberFlush(this);
+                if (_count < MaxBuffersPerArraySizePerCore)
+                {
+                    _arrays[_count++] = array;
+                    enqueued = true;
+                }
+                Monitor.Exit(this);
+                return enqueued;
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public T[] TryPop()
+            {
+                T[] arr = null;
+                MonitorEnterWithProcNumberFlush(this);
+                if (_count > 0)
+                {
+                    arr = _arrays[--_count];
+                    _arrays[_count] = null;
+                }
+                Monitor.Exit(this);
+                return arr;
+            }
+
+            /// <summary>
+            /// Enters the monitor on the object.  If there is any contention while trying
+            /// to acquire the monitor, it flushes the cached processor number so that subsequent
+            /// attempts to access the per-core stacks will use an updated processor number.
+            /// </summary>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private static void MonitorEnterWithProcNumberFlush(object obj)
+            {
+                if (!Monitor.TryEnter(obj))
```

Fail fast? https://github.com/dotnet/coreclr/pull/6951

`if (!Monitor.TryEnter(obj, 0))`